IGCexpansion/CodonGeneconv.py

```diff
@@ -653,11 +686,32 @@ def loglikelihood_and_gradient(self, display = False):
                 if i in self.Force.keys():  # check here
                     other_derivs.append(0.0)
                     continue
+            # x_plus_delta = np.array(self.x)
+            # x_plus_delta[i] += delta
+            # self.update_by_x(x_plus_delta)
+            # ll_delta, _ = fn(store = True, edge_derivative = False)
+            # d_estimate = (ll_delta - ll) / delta
+
+            # finite difference central
+
+            delta = max(1, abs(self.x[i])) * 0.000001
+            # reference: http://paulklein.ca/newsite/teaching/Notes_NumericalDifferentiation.pdf
+            # line before equation 22.
+
+            # ll_delta_plus is f(x+2/h)
             x_plus_delta = np.array(self.x)
-            x_plus_delta[i] += delta
+            x_plus_delta[i] += delta / 2
             self.update_by_x(x_plus_delta)
-            ll_delta, _ = fn(store = True, edge_derivative = False)
-            d_estimate = (ll_delta - ll) / delta           
+            ll_delta_plus, _ = fn(store=True, edge_derivative=False)
+
+            # ll_delta_minus is f(x-2/h)
+            x_minus_delta = np.array(self.x)

```

Comment:
I think that `self.x` is changed at this point because of `self.update_by_x()` which means that you need to change the next line to `x_minus_delta[i] -= delta`.  Put a stop point here in PyCharm and confirm if this is true.

Reply:
It shows that delta won't change after monitoring with debugger.

Reply:
No.  I asked you to check if `-= delta / 2` get you to the correct `x` value.

Reply:
You don't need to change it if it is correct.

Reply:
I get the correct value when I use `x_minus_delta[I] -= delta`. I think now it is correct.

Reply:
Yes, that's the correct check.